Add option to namespace the metadata

adamkiss opened the pull request:
For one of my metalsmith installations, I'd like to have all files' metadata grouped under `page.*` (e.g. `page.title`).

This is a quick (and naive) implementation of this functionality to open communication about it. It passes tests, but it introduces first non-gray-matter option, so you might want me to implement it in different way (or add other tests, based on comments in `test/index.js`)

adamkiss committed:
Prepare test for group option

adamkiss committed:
Add naive group implementation

adamkiss committed:
opts.group: Add test for nonexistent key (non-grouped)

Ajedi32 commented:
Seems reasonable. Just a couple things:
1. I think perhaps a better name for this option would be `namespace`. Thoughts?
2. This feature needs to be documented in the README.

adamkiss commented:
To be honest, I haven't touched README yet, because I didn't think this was production ready; I wasn't sure if pulling what's implemented (using `if option === 'implementedOption`) and sending the rest to gray-matter is a good way.
- replacing `group` with `namespace` is okay with me.
- documentation: If everything else is okay — and the code is written well in your opinion, I can rename the stuff to 'namespace' and document it.

Ajedi32 commented:
Ah okay, I didn't realize this was still a WIP.

The general approach of pulling the `metalsmith-matters`-specific options out of the options argument and sending the rest to `gray-matter` seems reasonable to me. There are perhaps a few things that could be done to make the code look cleaner, but overall it looks pretty good.

Why? Did you have other ideas in mind for the API?

adamkiss commented:
I just sometimes overthink things :)

adamkiss committed:
Rename 'group' to 'namespace', update tests

adamkiss committed:
Use 'namespace' if only string is given add the test

adamkiss committed:
Update readme with the namespace option

adamkiss committed:
Remove computed property names, thats ES6

adamkiss commented:
Right, so I changed the option to `namespace`, added namespacing if only a string is passed as options, and update the readme.

If it's good for you, you can bump the version and release.

Ajedi32 commented on lib/index.js:
```diff
@@ -47,6 +47,8 @@ function extractFrontmatter(file, filePath, options, gm_options){
 function frontmatter(opts){
+  if (typeof opts === 'string') {
+    opts = { namespace: opts };
   }
```
I'm a little bit skeptical of whether or not this feature is a good idea. :confused: 

It just doesn't seem like very obvious behavior to me that a string passed directly to `metalsmith-matters` indicates the namespace. Why not use it to indicate language? Why not the parser? I think I'd rather keep this explicit and require named keys for each option.

Ajedi32 commented on lib/index.js:
```diff
@@ -37,12 +44,24 @@ function extractFrontmatter(file, filePath, options){
+function frontmatter(opts){
+  if (typeof opts === 'string') {
+    opts = { namespace: opts };
+  }
```
I'm a little bit skeptical of whether or not this feature is a good idea. :confused: 

It just doesn't seem like very obvious behavior to me that a string passed directly to `metalsmith-matters` indicates the namespace. Why not use it to indicate language? Why not the parser? I think I'd rather keep this explicit and require named keys for each option.

adamkiss replied:
The idea behind it was that it was the only parameter given to metalsmith-matters. 

But I'm okay if you remove it after pull. It was spur of the moment thing. Or should I remove it?

adamkiss replied:
And by 'only one given to' I meant the only one we don't pass through to graymatter.

Ajedi32 replied:
Yeah you can go ahead and remove it. Which options are passed to gray matter and which are processed directly in metalsmith-matters is an implementation detail users shouldn't need to worry about - it doesn't need to be reflected in the API.

adamkiss committed:
Remove option to namespace matter by passing just a string as 'options'

adamkiss commented:
Go ahead and release this :)

Ajedi32 commented:
There we go. Sorry it took me so long to get to this. I'll cut a new release now.
